### PR TITLE
Balances removal follow up

### PR DESCRIPTION
**Pull request by nebolax**
## Checklist

- [x] Remove return  from `on_account_addition` callbacks. Now these functions don't return anything
- [x] Change blockchain accounts endpoints' responses. Now account addition returns accounts info and account removal returns balances update

### CODE REVIEW

**Comment by lukicenturi** [comment]
Tried the PR, there are some comments:
image
1. This is the response from the account addition, don't you think the response is too deep nested?
2. For the deletion, what I need is all removed assets with zero balances. For example, I have 1 account with 10 assets, then if I delete this account, it will return 10 assets with zero balances. Can you do that, AFAIK we go this way before the query optimization.

File: docs/api.rst [hunk 1]
```diff
@@ -9053,7 +9053,7 @@ Adding blockchain accounts
 
       This endpoint can also be queried asynchronously by using ``"async_query": true``
```
Comment (yabirgb) [inline docs/api.rst:9055]: The documentation when an account is removed should be also be updated

**Comment by nebolax** [comment]
> 1. This is the response from the account addition, don't you think the response is too deep nested?

I also noticed it when was adjusting tests, but thought it's okay :+1:. Pushed the fix.


> 2. For the deletion, what I need is all removed assets with zero balances. For example, I have 1 account with 10 assets, then if I delete this account, it will return 10 assets with zero balances. Can you do that, as I remember we go this way before the query optimization.

Discussed in DMs with Luki and decided to keep as is (just mentioning for the future reviewers)

File: rotkehlchen/chain/ethereum/modules/eth2/eth2.py [hunk 2]
```diff
@@ -460,7 +460,7 @@ def add_validator(
         ])
 
     # -- Methods following the EthereumModule interface -- #
-    def on_account_addition(self, address: ChecksumEthAddress) -> Optional[List[AssetBalance]]:  # pylint: disable=useless-return  # noqa: E501
+    def on_account_addition(self, address: ChecksumEthAddress) -> None:  # pylint: disable=useless-return  # noqa: E501
         """Just query balances and add detected validators to DB. Return nothing"""
         try:
             self.get_balances(
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/ethereum/modules/eth2/eth2.py:466]: I think it should not be removed since the methods called here don't return anything but add some data to the decoder instance, or am I wrong?

File: rotkehlchen/chain/ethereum/modules/liquity/trove.py [hunk 3]
```diff
@@ -553,15 +553,3 @@ def get_staking_history(
                     )
                     continue
         return result
-
-    # -- Methods following the EthereumModule interface -- #
-    def on_account_addition(self, address: ChecksumEthAddress) -> Optional[List['AssetBalance']]:
-        super().on_account_addition(address)
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/ethereum/modules/liquity/trove.py:559]: Why? This is a parent's method so it will be called automatically

File: docs/api.rst [hunk 4]
```diff
@@ -9053,7 +9053,7 @@ Adding blockchain accounts
...
 
-   Doing a PUT on the the blockchains endpoint with a specific blockchain URL and a list of account data in the json data will add these accounts to the tracked accounts for the given blockchain and the current user. The updated balances after the account additions are returned.
+   Doing a PUT on the blockchains endpoint with a specific blockchain URL and a list of account data in the json data will add these accounts to the tracked accounts for the given blockchain and the current user. The updated balances after the account additions are returned.
```
[the remark this answers is not in the document]
Comment (nebolax) [reply docs/api.rst:9056]: Oh yeah, thanks

File: rotkehlchen/tests/api/test_blockchain.py [hunk 5]
```diff
@@ -1049,7 +1051,19 @@ def test_add_blockchain_accounts_with_tags_and_label_and_querying_them(rotkehlch
...
+    for account_data_entry in accounts_data:
+        for result_entry in result['result']:
+            if account_data_entry['address'] == result_entry['address']:
+                if 'label' in account_data_entry:
+                    assert result_entry['label'] == account_data_entry['label']
+                if 'tags' in account_data_entry:
+                    assert set(result_entry['tags']) == set(account_data_entry['tags'])
+                break  # break means that we found the account data entry
+        else:  # else means that `break` has not happened
+            raise AssertionError(f'Not found entry {account_data_entry} in returned result')
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/tests/api/test_blockchain.py]: No, I can't since the order of the accounts in the list might be different and the order of the tags in the tags list of a particular account can also be different

File: rotkehlchen/chain/ethereum/modules/liquity/trove.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/chain/ethereum/modules/liquity/trove.py:559]: Yes, true. Don't know why I was thinking that you were using pass here

File: docs/api.rst (hunk 1 above)
Comment (nebolax) [reply docs/api.rst:9055]: M? What's changed?

Comment (yabirgb) [reply docs/api.rst:9055]: You now are returning the balances right? This should be in the response at the docs

Comment (nebolax) [reply docs/api.rst:9055]: M? What's changed?

File: rotkehlchen/tests/api/test_blockchain.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/tests/api/test_blockchain.py]: Couldn't you make a set so the order doesn't matter and iterate the result and make a set out of the field that is needed in the response? Also are you sure that the order in the response is different each time?

File: docs/api.rst (hunk 1 above)
Comment (nebolax) [reply docs/api.rst:9055]: Well, I am returning balances but I haven't changed docs in the balances query removal pr :smiley:, so it is the same as before that pr (and thus says that balances update is returned)

File: rotkehlchen/chain/ethereum/modules/eth2/eth2.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/chain/ethereum/modules/eth2/eth2.py:466]: Read what it does. There is one part that queries the validators and adds them to the databasse https://github.com/rotki/rotki/blob/ae10d6b55f6fb6f75e6be8d231f22a422ef19b7f/rotkehlchen/chain/ethereum/modules/eth2/eth2.py#L196 but there is other that is querying beaconchain for balances https://github.com/rotki/rotki/blob/develop/rotkehlchen/chain/ethereum/modules/eth2/eth2.py#L217 . You don't need that

File: rotkehlchen/api/rest.py [hunk 6]
```diff
@@ -1964,14 +1966,16 @@ def _remove_blockchain_accounts(
...
+            balances_update = self.rotkehlchen.chain_manager.get_balances_update()
         except EthSyncError as e:
             return {'result': None, 'message': str(e), 'status_code': HTTPStatus.CONFLICT}
         except InputError as e:
             return {'result': None, 'message': str(e), 'status_code': HTTPStatus.BAD_REQUEST}
         except RemoteError as e:
             return {'result': None, 'message': str(e), 'status_code': HTTPStatus.BAD_GATEWAY}
 
-        return OK_RESULT
+        # success
```
Comment (yabirgb) [inline rotkehlchen/api/rest.py:1977]: Comment not needed

File: rotkehlchen/tests/api/test_blockchain.py [hunk 7]
```diff
@@ -1603,17 +1612,24 @@ def test_remove_blockchain_accounts(
...
-        task_id = assert_ok_async_response(response)
-        wait_for_async_task(rotkehlchen_api_server, task_id)
-    else:
-        assert_proper_response(response)
-    assert rotki.chain_manager.accounts.btc == btc_accounts_after_removal
+    with ExitStack() as stack:
+        setup.enter_blockchain_patches(stack)
+        response = requests.delete(api_url_for(
+            rotkehlchen_api_server,
+            "blockchainsaccountsresource",
```
Comment (yabirgb) [inline rotkehlchen/tests/api/test_blockchain.py:1619]:
```suggestion
            'blockchainsaccountsresource',
```

File: rotkehlchen/tests/api/test_blockchain.py [hunk 8]
```diff
@@ -1603,17 +1612,24 @@ def test_remove_blockchain_accounts(
...
+            blockchain='BTC',
```
Comment (yabirgb) [inline rotkehlchen/tests/api/test_blockchain.py:1620]: With lefteris you commented that was better to use the hardcoded value or the enum.value?

File: rotkehlchen/tests/api/test_blockchain.py [hunk 9]
```diff
@@ -2208,34 +2241,50 @@ def test_remove_avax_blockchain_account(rotkehlchen_api_server):
...
     async_query = random.choice([False, True])
-    response = requests.get(api_url_for(
+    setup = setup_balances(
+        rotki=rotki,
+        ethereum_accounts=None,
+        btc_accounts=None,
+        eth_balances=None,
+        token_balances=None,
+        btc_balances=None,
+    )
```
Comment (yabirgb) [inline rotkehlchen/tests/api/test_blockchain.py:2251]: Why setup the balances if you are only testing avalanche?

File: rotkehlchen/tests/api/test_blockchain.py [hunk 10]
```diff
@@ -1928,38 +1944,55 @@ def test_remove_ksm_blockchain_account(rotkehlchen_api_server):
...
```
Comment (yabirgb) [inline rotkehlchen/tests/api/test_blockchain.py:1954]: why adding balances if you are testing kusama?

File: rotkehlchen/tests/api/test_blockchain.py [hunk 11]
```diff
@@ -2208,34 +2241,50 @@ def test_remove_avax_blockchain_account(rotkehlchen_api_server):
...
+        )
+        if async_query:
+            task_id = assert_ok_async_response(response)
+            result = wait_for_async_task_with_result(rotkehlchen_api_server, task_id)
+        else:
+            result = assert_proper_response_with_result(response)
 
-    assert rotki.chain_manager.accounts.avax == [AVALANCHE_ACC1_AVAX_ADDR]
-    assert rotki.chain_manager.totals.assets[A_AVAX].amount < avax_totals_beginning  # check that totals were modified  # noqa: E501
-    assert rotki.chain_manager.balances.avax[AVALANCHE_ACC1_AVAX_ADDR] == rotki.chain_manager.totals  # noqa: E501
```
Comment (yabirgb) [inline rotkehlchen/tests/api/test_blockchain.py:2237]: why remove this two checks?

**Review by yabirgb — COMMENTED** [review]
Some questions and comments @nebolax

File: rotkehlchen/tests/api/test_blockchain.py (hunk 11 above)
Comment (nebolax) [reply rotkehlchen/tests/api/test_blockchain.py:2237]: Because we now check totals in response

File: rotkehlchen/chain/ethereum/modules/liquity/trove.py [hunk 12]
```diff
@@ -553,15 +553,3 @@ def get_staking_history(
                     )
                     continue
         return result
-
-    # -- Methods following the EthereumModule interface -- #
-    def on_account_addition(self, address: ChecksumEvmAddress) -> Optional[List['AssetBalance']]:
```
Comment (prettyirrelevant) [inline rotkehlchen/chain/ethereum/modules/liquity/trove.py:558]: In other places where this is done, I see this
```py
def on_account_addition(self, address: ChecksumEvmAddress) -> None:
    pass
```
Any reason that's not done here?

File: rotkehlchen/chain/manager.py [hunk 13]
```diff
@@ -984,6 +984,8 @@ def modify_blockchain_accounts(
         # we are adding/removing accounts, make sure query cache is flushed
         self.flush_cache('query_balances')
         self.flush_cache('query_balances', blockchain=blockchain)
+        self.flush_cache('query_balances', blockchain=None, ignore_cache=False, beaconchain_fetch_eth1=False)  # noqa: E501
+        self.flush_cache('query_balances', blockchain=blockchain, ignore_cache=False, beaconchain_fetch_eth1=False)  # noqa: E501
```
Comment (prettyirrelevant) [inline rotkehlchen/chain/manager.py:988]: Question: Why did you add this here(and every other location)?

This is me trying to understand why.

**Review by prettyirrelevant — COMMENTED** [review]
LGTM. Just two questions

File: rotkehlchen/chain/manager.py (hunk 13 above)
Comment (nebolax) [reply rotkehlchen/chain/manager.py:988]: This is due to the cache key is calculated from the function name + kwargs and these combinations of arguments might be passed from `rest.py` to the `query_balances` method, so we have to clear cache for these combinations

File: rotkehlchen/chain/ethereum/modules/liquity/trove.py (hunk 12 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/modules/liquity/trove.py:558]: Yep, the reason is that this decoder is inherited from `HasDSProxy` which implements this method, so no need to override it here

Comment (LefterisJP) [reply rotkehlchen/chain/ethereum/modules/liquity/trove.py:558]: This will be handled anyway via the balance query, later right?

File: rotkehlchen/chain/manager.py (hunk 13 above)
Comment (LefterisJP) [reply rotkehlchen/chain/manager.py:988]: Hmm interesting. Damn this is a really ugly thing we have here with these caches. Perhaps need to reconsider them.

File: rotkehlchen/tests/api/test_blockchain.py [hunk 14]
```diff
@@ -1825,35 +1830,26 @@ def test_add_btc_blockchain_account_ens_domain(rotkehlchen_api_server):
...
-        rotki=rotkehlchen_api_server.rest_api.rotkehlchen,
-        ethereum_accounts=None,
-        btc_accounts=None,
-        eth_balances=None,
-        token_balances=None,
-        btc_balances=None,
+    response = requests.put(
+        api_url_for(
+            rotkehlchen_api_server,
+            "blockchainsaccountsresource",
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_blockchain.py:1848]:
```suggestion
            'blockchainsaccountsresource',
```

File: rotkehlchen/tests/api/test_blockchain.py [hunk 15]
```diff
@@ -1871,32 +1867,23 @@ def test_add_ksm_blockchain_account(rotkehlchen_api_server):
...
-        ethereum_accounts=None,
-        btc_accounts=None,
-        eth_balances=None,
-        token_balances=None,
-        btc_balances=None,
+
+    response = requests.put(
+        api_url_for(
+            rotkehlchen_api_server,
+            "blockchainsaccountsresource",
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_blockchain.py:1874]:
```suggestion
            'blockchainsaccountsresource',
```

File: rotkehlchen/tests/api/test_blockchain.py [hunk 16]
```diff
@@ -1968,24 +1963,14 @@ def test_add_ksm_blockchain_account_invalid_ens_domain(rotkehlchen_api_server):
...
-        rotki=rotkehlchen_api_server.rest_api.rotkehlchen,
-        ethereum_accounts=None,
-        btc_accounts=None,
-        eth_balances=None,
-        token_balances=None,
-        btc_balances=None,
+    response = requests.put(
+        api_url_for(
+            rotkehlchen_api_server,
+            "blockchainsaccountsresource",
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_blockchain.py:1969]:
```suggestion
            'blockchainsaccountsresource',
```

File: rotkehlchen/tests/api/test_blockchain.py [hunk 17]
```diff
@@ -1928,38 +1915,46 @@ def test_remove_ksm_blockchain_account(rotkehlchen_api_server):
...
+    ))  # to populate balances
     response = requests.delete(
         api_url_for(
             rotkehlchen_api_server,
             'blockchainsaccountsresource',
             blockchain=SupportedBlockchain.KUSAMA.value,
         ),
         json={
-            'accounts': [SUBSTRATE_ACC1_KSM_ADDR],
+            'accounts': [SUBSTRATE_ACC2_KSM_ADDR],
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_blockchain.py:1930]: Why in this test the acc changed?

File: rotkehlchen/tests/api/test_blockchain.py [hunk 18]
```diff
@@ -2065,24 +2051,14 @@ def test_remove_ksm_blockchain_account_ens_domain(rotkehlchen_api_server):
...
-        rotki=rotkehlchen_api_server.rest_api.rotkehlchen,
-        ethereum_accounts=None,
-        btc_accounts=None,
-        eth_balances=None,
-        token_balances=None,
-        btc_balances=None,
+    response = requests.put(
+        api_url_for(
+            rotkehlchen_api_server,
+            "blockchainsaccountsresource",
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_blockchain.py:2057]:
```suggestion
            'blockchainsaccountsresource',
```

File: rotkehlchen/tests/api/test_blockchain.py [hunk 19]
```diff
@@ -2097,32 +2073,22 @@ def test_add_avax_blockchain_account(rotkehlchen_api_server):
...
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_blockchain.py:2079]:
```suggestion
            'blockchainsaccountsresource',
```

File: docs/api.rst [hunk 20]
```diff
@@ -9073,7 +9073,7 @@ Adding blockchain accounts
 
       This endpoint can also be queried asynchronously by using ``"async_query": true``
 
-   Doing a PUT on the the blockchains endpoint with a specific blockchain URL and a list of account data in the json data will add these accounts to the tracked accounts for the given blockchain and the current user. The updated balances after the account additions are returned.
+   Doing a PUT on the blockchains endpoint with a specific blockchain URL and a list of account data in the json data will add these accounts to the tracked accounts for the given blockchain and the current user. A list of accounts that are in rotki is returned.
```
[the remark this answers is not in the document]
Comment (nebolax) [reply docs/api.rst:9076]: Actually, none of these :) I updated the docs and the returned data is the list of accounts' addresses that were added during a request. (this list is returned in case caller added an ens name and needs it's resolved address for the further usage).

File: rotkehlchen/chain/ethereum/modules/liquity/trove.py (hunk 12 above)
Comment (nebolax) [reply rotkehlchen/chain/ethereum/modules/liquity/trove.py:558]: Yep, is handled [here](https://github.com/rotki/rotki/blob/develop/rotkehlchen/chain/manager.py#L1280) and [here](https://github.com/rotki/rotki/blob/develop/rotkehlchen/chain/manager.py#L1287)

File: rotkehlchen/chain/ethereum/modules/pickle_finance/main.py [hunk 21]
```diff
@@ -143,8 +143,8 @@ def balances_in_protocol(
         return balances_per_address
 
     # -- Methods following the EthereumModule interface -- #
-    def on_account_addition(self, address: ChecksumEvmAddress) -> Optional[List['AssetBalance']]:
-        return self.balances_in_protocol([address]).get(address, None)
+    def on_account_addition(self, address: ChecksumEvmAddress) -> None:
+        pass
```
[the remark this answers is not in the document]
Comment (nebolax) [reply rotkehlchen/chain/ethereum/modules/pickle_finance/main.py:147]: Yep, is handled [here](https://github.com/rotki/rotki/blob/develop/rotkehlchen/chain/manager.py#L1270)

File: rotkehlchen/tests/api/test_blockchain.py (hunk 17 above)
Comment (nebolax) [reply rotkehlchen/tests/api/test_blockchain.py:1930]: I actually just reverted this test to how it looked like before the balances query removal pr since balances update is now returned again.

Iirc I had to modify the account in that pr due to one of them had no balance but we had to check that totals change after the account removal.

File: docs/api.rst [hunk 22]
```diff
@@ -9073,7 +9073,7 @@ Adding blockchain accounts
 
       This endpoint can also be queried asynchronously by using ``"async_query": true``
 
-   Doing a PUT on the the blockchains endpoint with a specific blockchain URL and a list of account data in the json data will add these accounts to the tracked accounts for the given blockchain and the current user. The updated balances after the account additions are returned.
+   Doing a PUT on the blockchains endpoint with a specific blockchain URL and a list of account data in the json data will add these accounts to the tracked accounts for the given blockchain and the current user. A list of accounts' addresses that were added during a request is returned. This data is returned so that if you add an ens name, you get its name's resolved address for the further usage.
```
Comment (LefterisJP) [inline docs/api.rst:9076]:
```suggestion
   Doing a PUT on the blockchains endpoint with a specific blockchain URL and a list of account data in the json data will add these accounts to the tracked accounts for the given blockchain and the current user. A list of accounts' addresses that were added during a request is returned. This data is returned so that if you add an ens name, you get its name's resolved address for further usage.
```

File: rotkehlchen/tests/api/test_blockchain.py (hunk 17 above)
Comment (LefterisJP) [reply rotkehlchen/tests/api/test_blockchain.py:1930]: I see. Hmm I can't check the logic and test addy easily. Will just trust you on this.

**Review by LefterisJP — COMMENTED** [review]
LGTM.

Things to check in the future:
1. Small typo in the api docs I pointed out but won't stop the merging for this.

The test that failed it a covalent failure.

**Review by LefterisJP — APPROVED** [review]
LGTM
